lib/icinga/dependency.cpp

```diff
@@ -9,106 +9,170 @@
+ */
+class DependencyCycleChecker
 {
-	if (graph.Nodes[parent].OnStack) {
-		std::ostringstream oss;
-		oss << "Dependency cycle:\n";
-
-		for (auto& frame : graph.Stack) {
-			oss << frame.Node->GetReflectionType()->GetName() << " '" << frame.Node->GetName() << "'";
+	struct Node
+	{
+		bool Visited = false;
+		bool OnStack = false;
+		std::vector<Dependency::Ptr> ExtraDependencies;
+	};
+
+	// std::hash is only implemented starting from Boost 1.74, use a (ordered) map as fallback.
+	// https://github.com/boostorg/smart_ptr/commit/5a18ffdc5609a0e64b63e47cb81c4f0847e0c087
+#if BOOST_VERSION >= 107400
+	std::unordered_map<Checkable::Ptr, Node> m_Nodes;
+#else
+	std::map<Checkable::Ptr, Node> m_Nodes;
+#endif
+
+	// Stack representing the path currently visited by AssertNoCycle(). Dependency::Ptr represents an edge from its
+	// child to parent, Service::Ptr represents the implicit dependency of that service to its host.
+	std::vector<std::variant<Dependency::Ptr, Service::Ptr>> m_Stack;
+
+public:
+	/**
+	 * Add a dependency to this DependencyCycleChecker that will be considered by AssertNoCycle() in addition to
+	 * dependencies already registered to the checkables. This allows checking if additional dependencies would cause
+	 * a cycle before actually registering them to the checkables.
+	 *
+	 * @param dependency Dependency to additionally consider during the cycle search.
+	 */
+	void AddExtraDependency(Dependency::Ptr dependency)
+	{
+		auto& node = m_Nodes[dependency->GetChild()];
+		node.ExtraDependencies.push_back(std::move(dependency));
+	}
 
-			if (frame.Implicit) {
-				oss << " (implicit)";
+	/**
+	 * Searches the dependency graph for cycles and throws an exception if one is found.
+	 *
+	 * Only the part of the graph that's reachable from the starting node when traversing dependencies towards the
+	 * parents is searched. In order to check that there are no cycles in the whole dependency graph, this method
+	 * has to be called for every checkable. For this, the method can be called on the same DependencyCycleChecker
+	 * instance multiple times, in which case parts of the graph aren't searched twice. However, if the graph structure
+	 * changes, a new DependencyCycleChecker instance must be used.
+	 *
+	 * @param checkable Starting node for the cycle search.
+	 * @throws ScriptError A dependency cycle was found.
+	 */
+	void AssertNoCycle(const Checkable::Ptr& checkable)
+	{
+		auto& node = m_Nodes[checkable];
+
+		Log(LogInformation, "DependencyCycleChecker") << m_Stack.size() << " " << checkable->GetName();
+
+		if (node.OnStack) {
+			std::ostringstream s;
+			s << "Dependency cycle:";
+			for (const auto& obj : m_Stack) {
+				Checkable::Ptr child, parent;
+				Dependency::Ptr dependency;
+
+				if (std::holds_alternative<Dependency::Ptr>(obj)) {
+					dependency = std::get<Dependency::Ptr>(obj);
+					parent = dependency->GetParent();
+					child = dependency->GetChild();
```

Comment:
How about checking whether `parent` and `child` are the same here and rewording the error message accordingly from:
```bash
[2025-03-07 11:22:41 +0100] critical/config: Error: Dependency cycle:
	- Host "HostA" depends on Host "HostA" (Dependency "depa" in dependency-apply.conf: 76:1-76:26)
...
```

To something like: `Host "HostA" depends on itself (Dependency "depa" in dependency-apply.conf: 76:1-76:26)`

I think that would make it more obvious that no other host is involved in that dependency cycle.

Reply:
As I only really noticed this during testing of the PR, what do you think of also addressing the following properly? I mean if we're going over the whole error message anyways. :sweat_smile: 

> Note: I tested this by including the config in my test cluster, so this shows an unrelated host test-sat-a-231 that's also present there. It's part of the path that found the cycle but not actually part of the cycle itself (the path shown has a shape like the Greek letter ρ). The test below doesn't show that extra host, but that's only because the starting nodes for the search are picked differently, there was no change in the code that would prevent such a path from appearing.

Reply:
> what do you think of also addressing the following properly?

Well, if you can get it fixed in a reasonable amount of time, then yes. I actually did read that note, but I thought it only appears when using the master branch, not using this PR as well.

Reply:
I mean it's not conceptually difficult, you just have to search the stack for the position of the currently visited checkable (that you'd otherwise put on the stack if it wasn't already there), just needs some fiddling around that the code stays nice.

Reply:
> but I thought it only appears when using the master branch, not using this PR as well.

Actually, it's a bit better with the current state of the PR compared to the master after all, given that I only start the search from parents, so for ρ-shaped path to appear in the error message, the path leading to the cycle needs to be longer than one edge. (Still, I'm currently failing to write a config that should trigger it with the current state of the PR...)

Reply:
> Still, I'm currently failing to write a config that should trigger it with the current state of the PR...

I would just take this as _stop messing with me_ 😆 and leave this PR as it is. If someone does trigger this case in production, we can then use their configuration to enhance this further, but for now I wouldn't waste any more time on it.

Reply:
I have no idea what I was doing in Friday :see_no_evil:

I just wanted to test the self-loop case right now and accidentally triggered the other case with it. All it takes is adding the following to the config snippet from the PR description:

```
object Dependency "dep-self" {
	parent_host_name = "cyclic-host-a"
	child_host_name = "cyclic-host-a"
}
```

```
[2025-03-10 12:25:59 +0100] critical/config: Error: Dependency cycle:
	- Service "cyclic-host-a!example-service" depends on Host "cyclic-host-a" (implicit)
	- Host "cyclic-host-a" depends on itself (Dependency "dep-self" in /etc/icinga2/zones.d/master/dependencies.conf: 20:1-20:28)
```

Reply:
> I just wanted to test the self-loop case right now and accidentally triggered the other case with it.

The output you posted above looks fine to me! The cycle check was initiated by `example-service` and while following the implicit dependency to its parent, a self-reference was detected, so I don't see anything wrong with that.